### DISCUSSION TITLE
Permit Constructor to pass Getter-Only Auto-Properties to methods with 'out', 'ref', or 'in' modifiers.

### DISCUSSION THREAD

**Discussion by Ken-Clement** [dotnet/csharplang#2695] [Language Ideas]
Constructors allow Getter-Only Auto-Properties to be assigned.  This is explained [here](https://github.com/dotnet/csharplang/blob/master/spec/classes.md#automatically-implemented-properties) as:

> a getter-only auto-property can also be assigned to in the body of a constructor of the enclosing class. Such an assignment assigns directly to the readonly backing field of the property.

As a consequence, such a getter-only auto-property should also be usable in a method call as a parameter having an 'out', 'ref', or 'in' modifier.

Such a capability would be particularly useful when constructing an immutable object whose getter-only property values are computed in the constructor by a framework using fluent syntax with methods having 'out' parameters to capture intermediate results of the fluent computation.

This is possible now by using local variables to capture these 'out' values and assigning the local variables to the getter-only properties.  However, this is the very sort of boilerplate clutter that auto-properties were created to eliminate.

On the surface, this does not seem as though it would be particularly difficult to do and it would eliminate what is arguably an inconsistency in the language and make a desirable implementation pattern elegant.  I will examine the Roslyn source to validate this and even attempt such a modification.

**Comment by PathogenDavid** [comment]
Not necessarily opposed to this idea, but:

>  it would eliminate what is arguably an inconsistency in the language

This change would be trading one inconsistency for another. [You can't ever pass any properties by reference](https://sharplab.io/#v2:EYLgHgbALANALiAbgUwE4EsBmBPAPgAQGYACfAJmIBVkBnOAWACgBvJ490k/ARglKmIBlZHAAUAewCucYugB2MsAEo2HNQF4AfMTDF1xbgAZDAblXtznWQuIBZbAAVU4gA5o42YswDmIk8Ro/AF9LSyIqWjEVRjUASC0hEQlpO0dnN1QPJTNGEMYgA==).

Setting a get-only auto property in a constructor is implemented by the compiler as setting the backing field directly, so it should at least be somewhat simple to do what you're asking for.

--------------------------------

Here are some related suggestions to add this capability for properties in general:

https://github.com/dotnet/csharplang/issues/748
https://github.com/dotnet/csharplang/issues/1235
